# LDSC Pipeline
- This analysis was conducted in the LDSC environment.
  - https://github.com/bulik/ldsc
- We used three AD GWAS summary statistics.
  - Bellenguez, C. et al. New insights into the genetic etiology of alzheimer’s disease and related dementias. Nature genetics 54, 412–436 (2022).
  - Jansen, I. E. et al. Genome-wide meta-analysis identifies new loci and functional pathways influencing alzheimer’s disease risk. Nature genetics 51, 404–413 (2019).
  - Kunkle, B. W. et al. Genetic meta-analysis of diagnosed alzheimer’s disease identifies new risk loci and implicates aβ, tau, immunity and lipid processing. Nature genetics 51, 414–430 (2019).
- You can download the necessary LDSC-related files from the following websites.
  - https://zenodo.org/records/8292725
  - https://console.cloud.google.com/storage/browser/broad-alkesgroup-public-requester-pays/LDSCORE
## Quick Guide - The process of LDSC analysis
1. Run `ldsc_annot.sh`
2. Run `ldsc_main_loop.sh`
3. Run `ldsc_heritability_raw.sh`

## Details
### `ldsc_annot.sh`
- This file generates annotation files using significant peaks and 1000G EUR reference (HG38).

#### Inputs
  - BED file (peak information)
  - BIM file (1000G reference)
  - ANNOT path (output directory)
  
#### Checkpoints
  - To run this code,
    - Set all paths.
    - Check the selected methods for comparison (e.g., proposed, SCENT, and Signac).
    - Verify the BED file's name.
    - Check whether the output directory exists.
---
### `ldsc_main_loop.sh`
- This file runs LDSC analysis using 1000G EUR reference (HG38), annotations from target peaks (generated by `ldsc_annot.sh`), and a SNP list.

#### Inputs
  - BIM file (1000G reference)
  - ANNOT (annotations from the target peaks)
  - Output directory
  - SNPlist (HapMap3)
  
#### Checkpoints
  - To run this code,
    - Set all paths.
    - Check the selected methods for comparison (e.g., proposed, SCENT, and Signac).
    - Match the output file name with the annotation name for future analysis.
    - Check whether the output directory exists.

---
### `ldsc_heritability.sh`
- This file runs heritability analysis using summary statistics, target annotations, baselineLD annotations, weights, and SNP frequency.

#### Inputs
  - Summary statistics (GWAS result)
  - Specific cell-type annotations (e.g., Microglia)
  - Baseline annotations (e.g., BaselineLD-v.2.2)
  - Weights (from 1000G)
  - Frequency (from 1000G)
  - Output directory
  
#### Checkpoints
  - To run this code,
    - Set all paths.
    - Define the target summary statistics.
    - Define the summary statistics file name.
    - Check whether the output directory exists.

### `make_annot_ver1.py`
- This file is a slightly modified version of the original make_annot.py in LDSC, designed for saving annotations. It is used in `ldsc_annot.sh`.
