library(dplyr)
library(Seurat)
library(Matrix)
library(BSgenome.Hsapiens.UCSC.hg38)
library(sctransform)
source('sctransform_helper.R')
source('simulation_helper.R')
library("optparse")
option_list = list(
    make_option(c("--study"), type="character", default="brain_CG",
                help="which dataset to analyze", metavar="character"),
    make_option(c("--p_top_gene"), type="integer", default=2000,
                help="top genes to study", metavar="integer"),
    make_option(c("--p_top_peak"), type="integer", default=20000,
                help="top peaks to study", metavar="integer"), 
    make_option(c("--ct"), type="character", default="Microglia",
                help="which cell type", metavar="character"),
    make_option(c("--n_pairs"), type="integer", default=1000,
                help="number of randomly sampled pairs", metavar="integer"),
    make_option(c("--i_permu"), type="integer", default=1,
                help="which permutation replicate", metavar="integer"),
    make_option(c("--n_permu"), type="integer", default=100,
                help="number of permutations", metavar="integer"),
    make_option(c("--sampling"), type="character", default="random",
                help = "how to sample random pairs", metavar="character"),
    make_option(c("--min.cell"), type="integer", default=1,
                help="minimal number of counts to be considered", metavar="integer"),
    make_option(c("--mode"), type="character", default="permutation",
                help="what experiment to run", metavar="character"),
    make_option(c("--permu_peak"), type="character", default="TRUE",
                help="permute peak"),
    make_option(c("--sparse_factor"), type="numeric", default=1,
                help = "simulate peak data to have seq_depth=seq_depth*sparse_factor", metavar="numeric")
)
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
study <- opt$study
p_top_gene <- opt$p_top_gene
p_top_peak <- opt$p_top_peak
ct <- opt$ct
n_pairs <- opt$n_pairs
i_permu <- opt$i_permu
n_permu <- opt$n_permu
sampling <- opt$sampling
min.cell <- opt$min.cell
mode <- opt$mode
permu_peak <- (opt$permu_peak == 'TRUE')
sparse_factor <- opt$sparse_factor
sparse_factor_suffix <- ifelse(sparse_factor == 1, '', sprintf('s*%.1f_', sparse_factor))

print(sampling)
print(ct)
print(min.cell)
print(mode)

data_dir <- '/projects/chang/scGRN/peak_gene'
pre_dir <- sprintf('../preprocessing/output/%s/%s', study, ct)
# directory to load permuted data
permu_output_dir <- sprintf('output/%s/%s/%s',
                            ifelse(grepl('batch', mode), 
                                   'permutation_batch_effect', 'permutation'),
                            study, ct)
output_dir <- sprintf('output/%s/%s/%s', mode, study, ct)
obs_dir <- sprintf('../real_data/output/%s/%s', study, ct)

# load data
if(study == 'brain_CG'){
    pn <- 'peaks'
    obj_suffix <- '_control'
}else if(grepl('PBMC', study)){
    pn <- 'peaks_ct'
    obj_suffix <- ''
}

print(Sys.time())

# load data - generated by preprocessing/save_seurat_obj_by_ct.R
ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, obj_suffix))

# evaluate SCTransform residuals to compute Signac correlations
#set.seed(2024)
#ct_obj <- Seurat::SCTransform(ct_obj, vst.flavor="v2")

# get DNA sequence information for peaks
#ct_obj[[pn]] <- RegionStats(
#    object = ct_obj[[pn]],
#    genome = BSgenome.Hsapiens.UCSC.hg38,
#    sep = c(":", "-")
#)
peak.mdata <- GetAssayData(object = ct_obj, assay = pn, layer = 'meta.features')

# load gene location
source(sprintf('../preprocessing/peak_annotation_helper.R'))
gene_promoters <- load.promoters()

# genes and peaks to focus on  (generated by preprocessing/get_cis_peak_gene_pairs.R)
top_genes <- read.table(sprintf('%s_top_%i_genes.txt', pre_dir, p_top_gene))[[1]]
top_peaks <- sprintf('%s_top_%i_peaks.txt', pre_dir, p_top_peak) %>% read.table
peak_gene_pairs <- sprintf('%s_%i_peak_%i_gene_pairs.txt', pre_dir, p_top_peak, p_top_gene) %>% read.table(header = T)

# Run Signac 
# based on an customized implementation of Signac LinkPeaks() in Signac_LinkPeaks.R
source('../method_implementation/Signac_LinkPeaks.R')

####
# run Signac on the observed data to save background mean and sd
####
obs_fn <- sprintf('%s_Signac_observed_results.txt', obs_dir)
if(!file.exists(obs_fn)){
    print('run Signac on observed data')
    res_list <- list()
    for(i in 1:p_top_gene){
	print(i)
        g <- top_genes[i]
        peak_inds <- peak_gene_pairs$gene == g
        promoter_inds <- gene_promoters$gene == g
	if(!any(peak_inds)) next()
        res_list[[i]] <- Signac_LinkPeaks(
            peak_gene_pairs$peak[peak_inds], # peaks to be tested for a gene
            seqnames(gene_promoters)[promoter_inds] %>% as.character, # which chromosome the selected gene is on, format: chr10
            t(x = ct_obj[['SCT']]@data[g, , drop = FALSE]), # normalized gene expression from sctransform slot data
            ct_obj, # seurat object for a cell type
	    peak.assay = pn,
	    seed = 1,
            min.cell = 1
        )
        res_list[[i]]$gene <- g
    }
    res_df <- do.call(rbind, res_list)
    colnames(res_df)[colnames(res_df) == 'peaks'] <- 'peak'
    write.table(res_df, obs_fn, quote = F)
}else{
    print('load Signac estimates on real data')
    res_df <- read.table(obs_fn)
}

####
# run Signac on randomly permuted data
####
sampled_pairs <- sprintf('%s_RNA_%i_pairs_%s_sampled_pairs.txt', permu_output_dir, n_pairs, sampling) %>% read.table
sampled_pairs$pair <- paste(sampled_pairs$peak, sampled_pairs$gene, sep = '|')
res_df$pair <- paste(res_df$peak, res_df$gene, sep = '|')
sampled_res_df <- res_df[match(sampled_pairs$pair, res_df$pair),]

print(sprintf('Start permu replicate %i', i_permu))
#for(i_permu in 1){
    start <- Sys.time()
    print(start)
    # load simulated data
    dat_list <- load_simulated_data(mode, 'Signac', i_permu, permu_peak, sparse_factor = sparse_factor)
    rna_counts <- dat_list$gene
    peak_counts <- dat_list$peak

    # if simulated RNA data will be used
    if(mode %in% c('permutation', 'permutation_batch_effect',
                   'simulation', 'simulation_batch_effect', 'simulation_null_batch_effect')){
        # https://github.com/satijalab/seurat/issues/5499
        reference.SCT.model <- ct_obj@assays$SCT@SCTModel.list$model1
        #rna_sct <- matrix(nrow = n_pairs, ncol = nrow(permu_rna_counts))
        #for(i_g in 1:n_pairs){
        #    rna_sct[i_g,] <- correct_counts(SCTModel_to_vst(SCTModel = reference.SCT.model), 
        #                                      rna_counts[i_g,,drop=F], verbosity=0)[1,]
        #}
        rna_sct <- correct_counts(SCTModel_to_vst(SCTModel = reference.SCT.model), 
                                  rna_counts, verbosity=0)
        # https://satijalab.org/seurat/reference/sctransform
        # "counts being (corrected) counts", "data being log1p(counts)"
        rna_sct <- log1p(rna_sct)
        # equivalent to 
        # tmp <- permu_ct_obj[['RNA']]@counts %>% as.matrix
        # tmp[sampled_pairs$gene,] <- t(permu_rna_counts)
        # permu_ct_obj[['RNA']]@counts <- Matrix(tmp, sparse = T)
        # permu_ct_obj <- Seurat::SCTransform(permu_ct_obj, vst.flavor="v2")
        # permu_sct_2 <- permu_ct_obj[['SCT']]@data
    }else{
        rna_sct <- ct_obj[['SCT']]@data[sampled_pairs$gene,]
    }

    # compute the raw correlations based sctransform residuals and peak counts
    # to be adjusted by background mean and sd computed from observed data
    cor_vec <- numeric(n_pairs)
    for(i_g in 1:n_pairs){
        cor_vec[i_g] <- qlcMatrix::corSparse(
          X = t(x = peak_counts[sampled_pairs$peak[i_g], , drop = FALSE]),
          Y = t(x = rna_sct[sampled_pairs$gene[i_g], , drop = FALSE])
        )
    }
    ts <- (cor_vec - sampled_res_df$bg.mean) / sampled_res_df$bg.sd
    res_mat <- matrix(c(cor_vec, ts), ncol = 2, byrow = F)
    end <- Sys.time()
    print(end)
    print(end - start)
    fn <-  sprintf('%s_Signac_%s_sampled_%si_permu_%i_results.rds', output_dir, sampling, sparse_factor_suffix, i_permu)
    saveRDS(res_mat, fn)
#}

    print(fn)

