library(dplyr)
library(Seurat)
library(MASS)
library(Matrix)

library("optparse")
option_list = list(
    make_option(c("--study"), type="character", default="brain_CG",
              help="which dataset to analyze", metavar="character"),
    make_option(c("--p_top_gene"), type="integer", default=2000,
              help="top genes to study", metavar="integer"),
    make_option(c("--p_top_peak"), type="integer", default=20000,
              help="top peaks to study", metavar="integer"), 
    make_option(c("--ct"), type="character", default="Microglia",
              help="which cell type", metavar="character"),
    #make_option(c("--i_ct"), type="integer", default=1,
	#	help="cell type index i", metavar="integer"),
    make_option(c("--n_pairs"), type="integer", default=1000,
              help="number of randomly sampled pairs", metavar="integer"),
    make_option(c("--n_simu"), type="integer", default=1,
              help="number of simulation replicates", metavar="integer"),
    make_option(c("--sampling"), type="character", default="random",
		help = "how to sample random pairs", metavar="character"),
    make_option(c("--simulation_mode"), type="character", default="realistic",
		help = "simulation mode", metavar="character"),
    make_option(c("--sparse_factor"), type="numeric", default=1,
                help = "simulate peak data to have seq_depth=seq_depth*sparse_factor", metavar="numeric")
)
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
study <- opt$study
p_top_gene <- opt$p_top_gene
p_top_peak <- opt$p_top_peak
ct <- opt$ct
#i_ct <- opt$i_ct
n_pairs <- opt$n_pairs
n_simu <- opt$n_simu
sampling <- opt$sampling
simulation_mode <- opt$simulation_mode
sparse_factor <- opt$sparse_factor
sparse_factor_suffix <- ifelse(sparse_factor == 1, '', sprintf('s*%.1f_', sparse_factor))
if(simulation_mode == 'realistic'){
    simu_str <- 'simulation'
}else if(simulation_mode == 'null'){
    simu_str <- 'simulation_null'
}else if(simulation_mode == 'null_permutation'){
    simu_str <- 'simulation_null_permutation'
}else if(simulation_mode == 'batch_effect'){
    simu_str <- 'simulation_batch_effect'
}else if(simulation_mode == 'null_batch_effect'){
    simu_str <- 'simulation_null_batch_effect'
}
# suffix for loading proposed estimates that adjusted for batch effects
batch_suffix <- ifelse(grepl('batch_effect', simulation_mode), '_batch_adjusted', '')
    
data_dir <- '/projects/chang/scGRN/peak_gene'


# load real data to extract sequencing depths and sample size
# load data
if(study == 'brain_CG'){
    pn <- 'peaks'
    obj_suffix <- '_control'
    cts <- c('Microglia', 'Excitatory', 'Inhibitory', 'Oligodendrocytes', 'Astrocytes')
}else if(grepl('PBMC', study)){
    pn <- 'peaks_ct'
    obj_suffix <- ''
    cts <- c('CD14 Mono', 'CD4 TCM', 'CD8 Naive', 'CD4 Naive', 'CD8 TEM', 'CD16 Mono')
}
#ct <- cts[i_ct]
print(ct)


pre_dir <- sprintf('../preprocessing/output/%s/%s', study, ct)
real_dir <- sprintf('../real_data/output/%s/%s', study, ct)
permu_output_dir <- sprintf('output/permutation/%s/%s', study, ct)
output_dir <- sprintf('output/%s/%s/%s', simu_str, study, ct)

# load data - generated by preprocessing/save_seurat_obj_by_ct.R
ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, obj_suffix))
rna_seq_depths <- colSums(ct_obj[['RNA']]@counts)
peak_seq_depths <- colSums(ct_obj[[pn]]@counts)
n_cells <- ncol(ct_obj[['RNA']]@counts)

# load sampled peaks 
peak_gene_pairs <- sprintf('%s_%i_peak_%i_gene_pairs.txt', pre_dir, p_top_peak, p_top_gene) %>% read.table(header = T)
sampled_pairs <- sprintf('%s_RNA_%i_pairs_%s_sampled_pairs.txt', permu_output_dir, n_pairs, sampling) %>% read.table(header = T)

# load marginal mean and variance previously estimated
if(simulation_mode %in% c('batch_effect', 'null_batch_effect')){
    proposed_est <- sprintf('%s_proposed_observed_results%s.txt', real_dir, batch_suffix) %>% read.table(header = T)
    batch_list <- sprintf('%s_proposed_observed_batch%s.rds', real_dir, batch_suffix) %>% readRDS
}else{
    proposed_est <- sprintf('%s_proposed_observed_results.txt', real_dir) %>% read.table(header = T)
}
# matching between sampled pairs and precomputed estimates
merged_df <- merge(sampled_pairs, proposed_est, by = c('peak', 'gene'), sort = F)

# set true correlations to those estimated from real data
merged_df$rho <- merged_df$covar / sqrt(merged_df$sigma_sq_gene * merged_df$sigma_sq_peak)
print(summary(merged_df$rho))
merged_df$rho[abs(merged_df$test_stat) < 1.96] <- 0 # set the pairs with unadjusted p value > 0.05 to 0
merged_df$rho[abs(merged_df$rho) > 1] <- 0
merged_df$rho[is.na(merged_df$rho)] <- 0
print(mean(merged_df$rho!=0))
print(summary(merged_df$rho[merged_df$rho!=0]))

print(sprintf('%i negative sigma_sq_gene estimates', sum(merged_df$sigma_sq_gene < 0)))
print(sprintf('%i negative sigma_sq_peak estimates', sum(merged_df$sigma_sq_peak < 0)))
merged_df$sigma_sq_gene[merged_df$sigma_sq_gene < 0] <- min(merged_df$sigma_sq_gene[merged_df$sigma_sq_gene>0])
merged_df$sigma_sq_peak[merged_df$sigma_sq_peak < 0] <- min(merged_df$sigma_sq_peak[merged_df$sigma_sq_peak>0])

print(sprintf('%i negative mu_gene estimates', sum(merged_df$mu_gene < 0)))
print(sprintf('%i negative mu_peak estimates', sum(merged_df$mu_peak < 0)))

if(grepl('batch_effect', simulation_mode)){
	# randomly select correlated pairs
    set.seed(2024)
    merged_df$rho <- merged_df$rho[sample.int(nrow(merged_df), replace=F)]
}


write.table(merged_df, sprintf('%s_RNA_%i_pairs_%s_sampled_realistic_simulation_pars%s.txt', output_dir, n_pairs, sampling, batch_suffix))
# ordering of peaks and genes are the same as sampled_pairs
all(merged_df$peak == sampled_pairs$peak) %>% print
all(merged_df$gene == sampled_pairs$gene) %>% print




# simulate peak and gene counts
# with sequencing depths, number of cells from real data
# and gene mean and variance estimated by our proposed method on real data
source('simulation_helper-checkpoint.R')

if(simulation_mode %in% c('realistic', 'null', 'batch_effect', 'null_batch_effect')){
    sim_rna_counts <- sim_peak_counts <- array(NA, dim = c(n_simu, n_cells, n_pairs))
    set.seed(2024)
    if(simulation_mode %in% c('batch_effect', 'null_batch_effect')){
        gene_mu_inds <- grepl('mu_gene', colnames(merged_df))
        peak_mu_inds <- grepl('mu_peak', colnames(merged_df))
    }
    for(i_pair in 1:n_pairs){
        if(i_pair %% 100 == 0) print(i_pair)
        #start <- Sys.time()
        if(simulation_mode %in% c('null', 'realistic')){
            pair_stat <- merged_df[i_pair, c('mu_gene', 'mu_peak', 'sigma_sq_gene', 'sigma_sq_peak', 'rho')] %>% as.numeric
            tmp <- sim_exper(rho = ifelse(simulation_mode == 'null', 0, pair_stat[5]), n = n_cells,
                         s_gene = rna_seq_depths, s_peak = peak_seq_depths * sparse_factor, # Reviewer 2, sparser ATAC-seq data 
                         mu_vec = pair_stat[1:2], var_vec = pair_stat[3:4], 
                         n_reps = n_simu, seed = i_pair*2 + 300, 
                         copula = T)
        }else if(simulation_mode %in% c('batch_effect', 'null_batch_effect')){
            pair_stat <- merged_df[i_pair, c('sigma_sq_gene', 'sigma_sq_peak', 'rho')] %>% as.numeric
            mu_mat <- matrix(c(merged_df[i_pair, gene_mu_inds] %>% unlist, merged_df[i_pair, peak_mu_inds]) %>% unlist, nrow=2, byrow=T)
            if(i_pair == n_pairs) print(mu_mat)
    	    # normalize batch effects such that its magnitude is ~20% of the average underying mean
    	    mu_mean_vec <- rowMeans(mu_mat)
    	    if(i_pair == n_pairs) print(mu_mean_vec)
    	    mu_mat <- t(scale(t(mu_mat), center = T, scale = T)) / sqrt(length(batch_list) )
    	    mu_mat <- mu_mean_vec + mu_mat * mu_mean_vec * 0.2
    	    if(i_pair == n_pairs) print(mu_mat)
    	    if(i_pair == n_pairs) print(apply(mu_mat, 1, mean))
    	    # control the over-dispersion
    	    var_vec <- pair_stat[1:2]
    	    min_theta <- apply(mu_mat, 1, min)^2 / var_vec
    	    if(i_pair == n_pairs) print(min_theta)
    	    var_vec <- var_vec * min_theta # such that the largest over-dispersion (sigma^2/mu^2) across batches is 1
    
    	    colnames(mu_mat) <- names(batch_list)
                tmp <- sim_exper_with_batch_effects(rho = ifelse(simulation_mode == 'batch_effect', pair_stat[3], 0), n = n_cells,
                                                s_gene = rna_seq_depths, s_peak = peak_seq_depths, 
                                                mu_mat = mu_mat, var_vec = var_vec, 
                                                batch_inds_list = batch_list,
                                                n_reps = n_simu, seed = i_pair*2 + 300)
        }
        sim_rna_counts[,,i_pair] <- tmp[[1]]
        sim_peak_counts[,,i_pair] <- tmp[[2]]
    }

    print('start saving data')
    start <- Sys.time()
    for(i_simu in 1:n_simu){
	fn <- sprintf('%s_RNA_%i_pairs_%s_sampled_i_simu_%i.txt', output_dir, n_pairs, sampling, i_simu)
        if(!file.exists(fn)){
	    print(fn)
            write.table(sim_rna_counts[i_simu,,], fn, quote = F)
	}
	fn <- sprintf('%s_peak_%i_pairs_%s_sampled_%si_simu_%i.txt', output_dir, n_pairs, sampling, sparse_factor_suffix, i_simu)
	if(!file.exists(fn)){
            print(fn)
            write.table(sim_peak_counts[i_simu,,], fn, quote = F)
	}
        print(anyNA(sim_rna_counts[i_simu,,]))
        print(anyNA(sim_peak_counts[i_simu,,]))
    }
    
}else if(simulation_mode == 'null_permutation'){
    sim_rna_counts <- array(NA, dim = c(n_simu, n_cells, n_pairs))
    set.seed(2024)
    theta_vec <- merged_df$mu_gene^2 / merged_df$sigma_sq_gene
    sigma_vec <- merged_df$sigma_sq_gene / merged_df$mu_gene
    for(i_pair in 1:n_pairs){
        #print(i_pair)
        start <- Sys.time()
        for(i_simu in 1:n_simu){
            z_gene <- rgamma(n_cells, shape=theta_vec[i_pair], scale=sigma_vec[i_pair])
            x_gene <- rpois(n_cells, rna_seq_depths * z_gene)
            sim_rna_counts[i_simu,,i_pair] <- rpois(n_cells, (x_gene / rna_seq_depths)[sample.int(n_cells, replace = F)] * rna_seq_depths)
        }
        #end <- Sys.time()
        #print(end - start)
    }
    print('start saving data')
    start <- Sys.time()
    for(i_simu in 1:n_simu){
        tmp <- Matrix(t(sim_rna_counts[i_simu,,]), sparse = T)
        # https://stackoverflow.com/questions/17574099/in-r-how-do-you-write-a-sparse-matrix-to-a-file
        writeMM(tmp, sprintf('%s_peak_%i_pairs_%s_sampled_i_simu_%i.txt', output_dir, n_pairs, sampling, i_simu))
    }
}

# print(sprintf('%s_RNA_%i_pairs_%s_sampled_i_simu_%i.txt', output_dir, n_pairs, sampling, i_simu))

print('all data saved')
end <- Sys.time()
print(end - start)
