library(Matrix)
library(dplyr)
library(Seurat)
library(Signac)
library(JASPAR2020)
library(TFBSTools)
library(motifmatchr)
library(BSgenome.Hsapiens.UCSC.hg38)
library(gprofiler2)

source('reproducibility_helper.R')
source('../moment_regressions.R')
library("optparse")
option_list = list(
    make_option(c("--study"), type="character", default="brain_CG",
              help="which dataset to analyze", metavar="character"),
    make_option(c("--p_top_gene"), type="integer", default=2000,
              help="top genes to study", metavar="integer"),
    make_option(c("--p_top_peak"), type="integer", default=20000,
              help="top peaks to study", metavar="integer"), 
    make_option(c("--i_ct"), type="integer", default=1,
              help="which cell type", metavar="integer"),
    make_option(c("--adjust_for_batch"), type = "character", default="FALSE",
               help="whether to adjust for batch effects", metavar="character"),
    make_option(c("--p_cutoff"), type = "numeric", default=0.05,
               help="p value cutoff for constrcting trios", metavar="numeric")
)
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
study <- opt$study
p_top_gene <- opt$p_top_gene
p_top_peak <- opt$p_top_peak
i_ct <- opt$i_ct
adjust_for_batch <- opt$adjust_for_batch == 'TRUE'
trio_pval_cutoff_input <- opt$p_cutoff

print(p_top_gene)
print(p_top_peak)

# load data
# generated by preprocessing/save_seurat_obj_by_ct
if(grepl('brain', study)){
    pn <- 'peaks'
    if(study == 'brain_CG'){
        cts <- c('Microglia', 'Excitatory', 'Inhibitory', 'Oligodendrocytes', 'Astrocytes')
        if(adjust_for_batch) batch_var_name <- 'id'
    }else if(study == 'brain_ROSMAP'){
        cts <- c('Mic', 'Exc', 'Inh', 'Oli', 'Ast')
        if(adjust_for_batch) batch_var_name <- 'Sample'
    }
}else if(grepl('PBMC', study)){
    pn <- 'peaks_ct'
    obj_suffix <- ''
    cts <- c('CD14 Mono', 'CD4 TCM', 'CD8 Naive', 'CD4 Naive', 'CD8 TEM') 
    if(adjust_for_batch) batch_var_name <- 'dataset'
}else if(grepl('BMMC', study)){
    pn <- 'peaks'
    obj_suffix <- ''
    cts <- c('CD8+ T', 'CD14+ Mono', 'NK', 'CD4+ T activated', 'Naive CD20+ B', 'Erythroblast', 'CD4+ T naive')
}

ct <- cts[i_ct]
print(ct)
pre_dir <- sprintf('../preprocessing/output/%s/%s', study, ct)
output_dir <- sprintf('output/%s/%s', study, ct)
ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, ''))
    
pairs_suffix <- ifelse(p_top_gene == 5000, '_p_top_5000_50000', '')

# load inferred links
proposed_list <- list(control = load_proposed(ct, study, sprintf('%s_batch_adjusted', pairs_suffix))[[1]])
#,
#                     disease = load_proposed(ct, study, sprintf('_disease%s_batch_adjusted', pairs_suffix))[[1]])
groups <- names(proposed_list)
TF_coexp_list <- list()
TF_coexp_list[['control']] <- readRDS(sprintf('%s_TF_gene_coexp%s.rds', output_dir, 
                                             sprintf('%s_batch_adjusted', pairs_suffix)))
#TF_coexp_list[['disease']] <- readRDS(sprintf('%s_TF_gene_coexp%s.rds', output_dir, 
#                                              sprintf('_disease%s_batch_adjusted', pairs_suffix)))

TF_peak_list <- list()
TF_peak_list[['control']] <- readRDS(sprintf('%s_TF_peak_coexp%s.rds', output_dir, 
                                             sprintf('%s_batch_adjusted', pairs_suffix)))
#TF_peak_list[['disease']] <- readRDS(sprintf('%s_TF_peak_coexp%s.rds', output_dir, 
#                                              sprintf('_disease%s_batch_adjusted', pairs_suffix)))
    
# -
# 1.  Extract enriched TFs
# -
source('trio_analysis_helper.R')

if(grepl('PBMC', study)){
    DefaultAssay(ct_obj) <- 'peaks_ct'
    motif.object <- readRDS(sprintf('../preprocessing/output/%s/motif.object.rds', study))
}else if(grepl('brain', study)){
    motif.object <- readRDS('output/brain_CG/motif.object.rds')
}
Motifs(ct_obj) <- motif.object
# extract all TF names from JASPAR 2020 database
full_TF_names <- ConvertMotifID(ct_obj, id = colnames(motif.object@data))
TF_names <- intersect(full_TF_names, rownames(ct_obj[['RNA']]))

p_var <- 'pval'
p_cutoff <- 0.05
sig_inds_list <- lapply(groups, function(g) which(proposed_list[[g]][[p_var]] < p_cutoff))
names(sig_inds_list) <- groups
for(g in groups){
    print(length(sig_inds_list[[g]]))
}

TF_enrich_list <- lapply(groups, function(g){
    linked_peaks <- proposed_list[[g]]$peak[sig_inds_list[[g]]]
    sparsity <- TF_coexp_list[[g]][[3]]
    TF_motif_enrich(linked_peaks, sparsity, ct_obj)
})
names(TF_enrich_list) <- groups

# -
# 1.  Construct trios for each TF
# -
if(grepl('brain', study)){
    trio_pval_cutoff_vec <- c(0.001, 0.001, 0.05, 0.05, 0.1)
    names(trio_pval_cutoff_vec) <- c('Excitatory', 'Inhibitory', 'Oligodendrocytes', 'Astrocytes', 'Microglia')
    trio_pval_cutoff <- trio_pval_cutoff_vec[ct]
}else if(grepl('PBMC', study)){
    trio_pval_cutoff_vec <- c(0.001)
    names(trio_pval_cutoff_vec) <- c('CD14 Mono')
    trio_pval_cutoff <- trio_pval_cutoff_vec[ct]
}
                        
trio_pval_cutoff <- ifelse(trio_pval_cutoff_input==0, trio_pval_cutoff, trio_pval_cutoff_input) #trio_pval_cutoff_vec[ct]
trio_simul <- T

TF_trios_list <- list()
for(g in groups){
    # focus on TFs that are (1) enriched (2) highly expressed
    high_enriched_TFs <- TF_enrich_list[[g]]$motif.name[which(TF_enrich_list[[g]]$sparsity < 0.95)]
    
    TF_trios_df_list <- list()
    for(TF in high_enriched_TFs){
        TF_trios_df_list[[TF]] <- make_trio(TF, proposed_list[[g]], 
                                            TF_coexp_list[[g]]$pval,
                                            TF_coexp_list[[ct]][[g]]$coexp,
                                            TF_peak_list[[g]][[TF]],
                                            trio_pval_cutoff, simul = trio_simul,
                                           subset = T)
    }
    TF_trios_df <- do.call(rbind, TF_trios_df_list)
    TF_trios_df$sparsity <- TF_enrich_list[[g]]$sparsity[match(TF_trios_df$TF, TF_enrich_list[[g]]$motif.name)]
    # remove trios with TF=target genes
    TF_trios_df <- TF_trios_df[!(TF_trios_df$TF == TF_trios_df$gene),]
    TF_trios_list[[g]] <- TF_trios_df
    print(dim(TF_trios_df))
}

fn <- sprintf('output/%s/trios_%s%s%s.rds', 
              study,
               ct, 
               pairs_suffix,
               sprintf('_pval_cutoff_%.2e', trio_pval_cutoff))
print(fn)
saveRDS(TF_trios_list,
       fn)