library(dplyr)
library(Seurat)
library(gbm)
library("optparse")
option_list = list(
    make_option(c("--study"), type="character", default="brain_CG",
              help="which dataset to analyze", metavar="character"),
    make_option(c("--p_top_gene"), type="integer", default=2000,
              help="top genes to study", metavar="integer"),
    make_option(c("--p_top_peak"), type="integer", default=20000,
              help="top peaks to study", metavar="integer"), 
    make_option(c("--i_ct"), type="integer", default=1,
              help="which cell type", metavar="integer"),
    make_option(c("--cell_subset"), type="character", default="Control",
                help="which subset of cells to analyze", metavar="character")
)
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
study <- opt$study
p_top_gene <- opt$p_top_gene
p_top_peak <- opt$p_top_peak
i_ct <- opt$i_ct
cell_subset <- opt$cell_subset
adjust_for_batch <- F
run_disease <- F
print(p_top_gene)
print(p_top_peak)


# -
# load data - generated by ../preprocessing/save_seurat_obj_by_ct.R
# -
if(grepl('brain', study)){
    pn <- 'peaks'
    if(study == 'brain_CG'){
        cts <- c('Microglia', 'Excitatory', 'Inhibitory', 'Oligodendrocytes', 'Astrocytes')
        if(adjust_for_batch) batch_var_name <- 'id'
    }else if(study == 'brain_ROSMAP'){
        cts <- c('Mic', 'Exc', 'Inh', 'Oli', 'Ast')
        if(adjust_for_batch) batch_var_name <- 'Sample'
    }
    if(cell_subset == 'Control'){
        obj_suffix <- '_control'
    }else if(cell_subset == 'All'){
        obj_suffix <- ''
    }
}else if(grepl('PBMC', study)){
    pn <- 'peaks_ct'
    obj_suffix <- ''
    cts <- c('CD14 Mono', 'CD4 TCM', 'CD8 Naive', 'CD4 Naive', 'CD8 TEM') 
    if(adjust_for_batch) batch_var_name <- 'dataset'
}else if(grepl('BMMC', study)){
    pn <- 'peaks'
    obj_suffix <- ''
    cts <- c('CD8+ T', 'CD14+ Mono', 'NK', 'CD4+ T activated', 'Naive CD20+ B', 'Erythroblast', 'CD4+ T naive')
}

ct <- cts[i_ct]
print(ct)
pre_dir <- sprintf('../preprocessing/output/%s/%s', study, ct)
output_dir <- sprintf('output/%s/%s', study, ct)

    
# run_disease means that we estimate associations in the disease cells
# for the same peak-gene pairs considered for the control cells
if(run_disease){
    # load single cell object for all cells
    ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, ''))
    if(study == 'brain_CG') ct_obj <- subset(ct_obj, Status == 'AD')
}else{
    ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, obj_suffix))
}
    
# obtain sequencing depths
rna_seq_depths <- colSums(ct_obj[['RNA']]@counts)
peak_seq_depths <- colSums(ct_obj[[pn]]@counts)

peak_gene_pairs <- sprintf('%s_%i_peak_%i_gene_pairs.txt', pre_dir, p_top_peak, p_top_gene) %>% read.table(header = T)
unique_genes <- unique(peak_gene_pairs$gene)
unique_peaks <- unique(peak_gene_pairs$peak)
gene_pair_match <- match(peak_gene_pairs$gene, unique_genes)
peak_pair_match <- match(peak_gene_pairs$peak, unique_peaks)

# evaluate the proposed method on observed data
rna_counts <- ct_obj[['RNA']]@counts[unique_genes, ] %>% as.matrix %>% t
peak_counts <- ct_obj[[pn]]@counts[unique_peaks, ] %>% as.matrix %>% t
    

# -
# run GBM
# -
print(Sys.time())
res_list <- list()
for(g in unique_genes){
    g_peaks <- peak_gene_pairs$peak[peak_gene_pairs$gene == g]
    df <- as.data.frame(cbind(gene = rna_counts[,g], peak_counts[,g_peaks]))
    if(length(g_peaks) == 1) colnames(df)[2] <- g_peaks
    gbm_fit <- gbm(gene ~ ., distribution = 'gaussian', data=df)
    gbm_fit_df <- summary.gbm(gbm_fit, plotit=F) # peaks in this df are ordered by rel.inf
    res_list[[g]] <- cbind(gbm_fit_df, gene = g)
}
print(Sys.time())
res_df <- do.call(rbind, res_list)
res_df$var <- gsub('`', '', res_df$var)
colnames(res_df)[1] <- 'peak'
    
saveRDS(res_df, 
        sprintf('output/%s/GBM_%s.rds', study, ct))