library(dplyr)
library(Seurat)
library(Matrix)
library(BSgenome.Hsapiens.UCSC.hg38)
library(sctransform)
library(optparse)
option_list = list(
    make_option(c("--study"), type="character", default="brain_CG",
              help="which dataset to analyze", metavar="character"),
    make_option(c("--p_top_gene"), type="integer", default=2000,
              help="top genes to study", metavar="integer"),
    make_option(c("--p_top_peak"), type="integer", default=20000,
              help="top peaks to study", metavar="integer"), 
    make_option(c("--i_ct"), type="integer", default=1,
              help="which cell type", metavar="integer"),
    make_option(c("--min.cell"), type="integer", default=10,
              help="minimal number of counts to be considered", metavar="integer"),
    make_option(c("--cell_subset"), type="character", default="Control",
                help="which subset of cells to analyze", metavar="character"),
    make_option(c("--permu"), type="character", default="FALSE",
		help="whether to permute the data"),
    make_option(c("--permu_within_batch"), type="character", default="FALSE",
		help="whether to perform the experiment to permute peak counts within batch"),
    make_option(c("--i_permu"), type="integer", default=1,
              help="which permutation replicate", metavar="integer"),
    make_option(c("--run_TF_peak"), type="character", default="FALSE",
                help="whether to run TF peak association")
)
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
study <- opt$study
p_top_gene <- opt$p_top_gene
p_top_peak <- opt$p_top_peak
i_ct <- opt$i_ct
min.cell <- opt$min.cell
cell_subset <- opt$cell_subset
i_permu <- opt$i_permu
run_TF_peak <- (opt$run_TF_peak == 'TRUE')
fn_suffix <- ifelse(cell_subset == 'Control', '', '_All_cells')
if(p_top_gene != 2000){
    fn_suffix <- paste0(fn_suffix, sprintf('_p_top_%i_%i', p_top_gene, p_top_peak))
}
print(cell_subset)
print(fn_suffix)

print('!!! min.cell !!!')
print(min.cell)

permu <- (opt$permu == 'TRUE')
permu_within_batch <- (opt$permu_within_batch == 'TRUE')

# -
# load data - generated by ../preprocessing/save_seurat_obj_by_ct.R
# -
if(grepl('brain', study)){
    pn <- 'peaks'
    if(study == 'brain_CG'){
        cts <- c('Microglia', 'Excitatory', 'Inhibitory', 'Oligodendrocytes', 'Astrocytes')
        if(permu_within_batch) batch_var_name <- 'id'
    }else if(study == 'brain_ROSMAP'){
        cts <- c('Mic', 'Exc', 'Inh', 'Oli', 'Ast')
        if(permu_within_batch) batch_var_name <- 'Sample'
    }
    if(cell_subset == 'Control'){
        obj_suffix <- '_control'
    }else if(cell_subset == 'All'){
        obj_suffix <- ''
    }
}else if(grepl('PBMC', study)){
    pn <- 'peaks_ct'
    obj_suffix <- ''
    cts <- c('CD14 Mono', 'CD4 TCM', 'CD8 Naive', 'CD4 Naive', 'CD8 TEM')
    if(permu_within_batch) batch_var_name <- 'dataset'
}
ct <- cts[i_ct]
print(ct)
pre_dir <- sprintf('../preprocessing/output/%s/%s', study, ct)
output_dir <- sprintf('output/%s/%s', study, ct)


ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, obj_suffix))

# load data on peak information and gene location
peak.mdata <- GetAssayData(object = ct_obj, assay = pn, layer = 'meta.features')
source('peak_annotation_helper.R')
gene_promoters <- load.promoters()

# genes and peaks to focus on  (generated by preprocessing/get_cis_peak_gene_pairs.R)
top_genes <- read.table(sprintf('%s_top_%i_genes.txt', pre_dir, p_top_gene))[[1]]
top_peaks <- read.table(sprintf('%s_top_%i_peaks.txt', pre_dir, p_top_peak))[[1]]
peak_gene_pairs <- sprintf('%s_%i_peak_%i_gene_pairs.txt', pre_dir, p_top_peak, p_top_gene) %>% read.table(header = T)

# Run Signac 
# based on an customized implementation of Signac LinkPeaks() in Signac_LinkPeaks.R
# that fix a bug and improve the speed
source('../method_implementation/Signac_LinkPeaks.R')

# run Signac on the observed data to save background mean and sd
source('../simulation/adjustment_helper-checkpoint.R')
rna_seq_depths <- colSums(ct_obj[['RNA']]@counts)
if(permu_within_batch | permu){ 
    # for p-value adjustment based on permutation
    print(sprintf('permute peak accessibility %s, i_permu=%i', ifelse(permu_within_batch, 'within batch', ''), i_permu))
    set.seed(2024)
    random_seeds <- sample(1:1000000, 100, replace = F)
    # save to dense matrix to modify the elements
    full_counts <- ct_obj[['RNA']]@counts %>% as.matrix 
    # permute the peak accessibilities
    if(permu_within_batch){
        batch_var <- ct_obj[[batch_var_name]][[1]]
        permu_counts <- permute_within_batches(t(full_counts[top_genes,]), rna_seq_depths, batch_var, random_seeds[i_permu])
    }else if(permu){
        permu_counts <- permute_expression(t(full_counts[top_genes,]), rna_seq_depths, random_seeds[i_permu])
    }
    # replace the observed values with permuted counts
    full_counts[top_genes,] <- t(permu_counts)
    ct_obj[['RNA']]@counts <- Matrix(full_counts, sparse = TRUE)
    obs_fn <- sprintf('%s_Signac_observed_results_%s_i_%i%s.txt', 
                      output_dir, 
                      ifelse(permu_within_batch, 'permu_within_batch', 'permu'),
                      i_permu, fn_suffix)
}else{
    obs_fn <- sprintf('%s_Signac_observed_results%s.txt', output_dir, fn_suffix)
    if(min.cell == 800) obs_fn <- sprintf('%s_Signac_observed_results_min.cell_800%s.txt', output_dir, fn_suffix)
}

# re-compute SCT with permuted data
set.seed(2024)
ct_obj <- Seurat::SCTransform(ct_obj, vst.flavor="v2", )

if(!file.exists(obs_fn)){
    start <- Sys.time()
    print(start)
    print(sprintf('run Signac on observed data, permu_within_batch=%s', opt$permu_within_batch))
    res_list <- list()
    for(i in 1:p_top_gene){
    	if(i%%100 == 0) print(i)
        print(i)    
        g <- top_genes[i]
        peak_inds <- peak_gene_pairs$gene == g
        promoter_inds <- gene_promoters$gene == g
    	if(!any(peak_inds)) next()
    	if(g %in% rownames(ct_obj[['SCT']]@data)){
    	    ge_sct <- ct_obj[['SCT']]@data[g, , drop = FALSE]
    	}else{
    	    # if the permuted gene expression was excluded from sctransform due to being classified as a Poisson gene
    	    ge_mean <- sum(ct_obj[['RNA']]@counts[g, , drop = FALSE]) / sum(rna_seq_depths)
            ge_sct <- (ct_obj[['RNA']]@counts[g, , drop = FALSE] - rna_seq_depths * ge_mean) / sqrt(rna_seq_depths * ge_mean)
    	}
        res_list[[i]] <- Signac_LinkPeaks(
            peak_gene_pairs$peak[peak_inds], # peaks to be tested for a gene
            seqnames(gene_promoters)[promoter_inds] %>% as.character, # which chromosome the selected gene is on, format: chr10
            t(x = ge_sct), # normalized gene expression from sctransform slot data
            ct_obj, # seurat object for a cell type
    	    peak.assay = pn,
    	    seed = 1,
            min.cell = min.cell #1
        )
        res_list[[i]]$gene <- g
    }
    res_df <- do.call(rbind, res_list)
    colnames(res_df)[colnames(res_df) == 'peaks'] <- 'peak'
    write.table(res_df, obs_fn, quote = F)
    end <- Sys.time()
    print(end)
    print(end - start)
}else{
	print('File exists')
}
print(obs_fn)
# check match with peak_gene_pairs


if(run_TF_peak & study == 'PBMC_4_combined' & ct == 'CD14 Mono'){
    # load candidate TF-peak pairs generated by get_trio_peak_TF_pairs.R
    TF_peak_pairs <- readRDS('output/PBMC_4_combined/trio_peak_TF_pairs.rds')[['Signac']]
    # for each TF, evaluate the association with peaks that harbor its binding motif
    res_list <- list()
    TF_names <- names(TF_peak_pairs)
    start_time <- Sys.time()
    for(i in 1:length(TF_peak_pairs)){
        g <- TF_names[i]
        if(length(TF_peak_pairs[[g]]) > 1){
            #
            # obtain gene expression for the TF
            #
            promoter_inds <- gene_promoters$gene == g
        	if(g %in% rownames(ct_obj[['SCT']]@data)){
        	    ge_sct <- ct_obj[['SCT']]@data[g, , drop = FALSE]
        	}else{
        	    # if the permuted gene expression was excluded from sctransform due to being classified as a Poisson gene
        	    ge_mean <- sum(ct_obj[['RNA']]@counts[g, , drop = FALSE]) / sum(rna_seq_depths)
                ge_sct <- (ct_obj[['RNA']]@counts[g, , drop = FALSE] - rna_seq_depths * ge_mean) / sqrt(rna_seq_depths * ge_mean)
        	}
            tmp <- Signac_LinkPeaks(
                TF_peak_pairs[[g]], # peaks to be tested for a gene
                seqnames(gene_promoters)[promoter_inds] %>% as.character, # which chromosome the selected gene is on, format: chr10
                t(x = ge_sct), # normalized gene expression from sctransform slot data
                ct_obj, # seurat object for a cell type
        	    peak.assay = pn,
        	    seed = 1,
                min.cell = min.cell #1
            )
            tmp$gene <- g
            rownames(tmp) <- TF_peak_pairs[[g]]
            res_list[[g]] <- tmp
        }
    }
    res_df <- do.call(rbind, res_list)
    colnames(res_df)[colnames(res_df) == 'peaks'] <- 'peak'
    
    #fn <- sprintf('%s_TF_peak_coexp%s%s.rds', output_dir, fn_suffix, batch_suffix
    fn_short <- strsplit(obs_fn, '.txt')[[1]][1]
    fn <- paste0(fn_short, '_TF_peak.txt')
    print(fn)
    if(length(res_list) < 10) print('Testing run_Signac.R and saving test results')
    write.table(res_df, fn, quote = F)
    print(Sys.time() - start_time)
}