library(dplyr)
library(Seurat)
library(scMultiMap)
source('../method_implementation/moment_regressions.R')
library(optparse)
option_list = list(
    make_option(c("--study"), type="character", default="brain_CG",
              help="which dataset to analyze", metavar="character"),
    make_option(c("--p_top_gene"), type="integer", default=2000,
              help="top genes to study", metavar="integer"),
    make_option(c("--p_top_peak"), type="integer", default=20000,
              help="top peaks to study", metavar="integer"), 
    make_option(c("--i_ct"), type="integer", default=1,
              help="which cell type", metavar="integer"),
    make_option(c("--cell_subset"), type="character", default="Control",
                help="which subset of cells to analyze", metavar="character"),
    make_option(c("--adjust_for_batch"), type = "character", default="FALSE",
               help="whether to adjust for batch effects", metavar="character"),
    make_option(c("--run_disease"), type = "character", default="FALSE",
               help="estimate peak-gene association for disease group", metavar="character"),
    make_option(c("--run_TF_gene"), type = "character", default="FALSE",
               help="estimate peak-gene association for TF and other genes", metavar="character")
)
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
study <- opt$study
p_top_gene <- opt$p_top_gene
p_top_peak <- opt$p_top_peak
i_ct <- opt$i_ct
cell_subset <- opt$cell_subset
adjust_for_batch <- opt$adjust_for_batch == 'TRUE'
run_disease <- opt$run_disease == 'TRUE'
run_TF_gene <- opt$run_TF_gene == 'TRUE'

print(p_top_gene)
print(p_top_peak)

# load data
# generated by preprocessing/save_seurat_obj_by_ct
if(grepl('brain', study)){
    pn <- 'peaks'
    if(study == 'brain_CG'){
        cts <- c('Microglia', 'Excitatory', 'Inhibitory', 'Oligodendrocytes', 'Astrocytes')
        if(adjust_for_batch) batch_var_name <- 'id'
    }else if(study == 'brain_ROSMAP'){
        cts <- c('Mic', 'Exc', 'Inh', 'Oli', 'Ast')
        if(adjust_for_batch) batch_var_name <- 'Sample'
    }
    if(cell_subset == 'Control'){
        obj_suffix <- '_control'
    }else if(cell_subset == 'All'){
        obj_suffix <- ''
    }
}else if(grepl('PBMC', study)){
    pn <- 'peaks_ct'
    obj_suffix <- ''
    cts <- c('CD14 Mono', 'CD4 TCM', 'CD8 Naive', 'CD4 Naive', 'CD8 TEM') 
    if(adjust_for_batch) batch_var_name <- 'dataset'
}else if(grepl('BMMC', study)){
    pn <- 'peaks'
    obj_suffix <- ''
    cts <- c('CD8+ T', 'CD14+ Mono', 'NK', 'CD4+ T activated', 'Naive CD20+ B', 'Erythroblast', 'CD4+ T naive')
}

ct <- cts[i_ct]
print(ct)
pre_dir <- sprintf('../preprocessing/output/%s/%s', study, ct)
output_dir <- sprintf('output/%s/%s', study, ct)

# run_disease means that we estimate associations in the disease cells
# for the same peak-gene pairs considered for the control cells
if(run_disease){
    # load single cell object for all cells
    ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, ''))
    if(study == 'brain_CG') ct_obj <- subset(ct_obj, Status == 'AD')
}else{
    ct_obj <- readRDS(sprintf('%s_seurat_obj%s.rds', pre_dir, obj_suffix))
}
    
# obtain file suffix
if(cell_subset == 'Control'){
    fn_suffix <- ifelse(run_disease, '_disease', '')
}else{
    fn_suffix <- '_All_cells'
}
if(p_top_gene != 2000){
    fn_suffix <- paste0(fn_suffix, sprintf('_p_top_%i_%i', p_top_gene, p_top_peak))
}
batch_suffix <- ifelse(adjust_for_batch, '_batch_adjusted', '')

# -
# run scMultiMap
# -
#peak_gene_pairs <- sprintf('%s_%i_peak_%i_gene_pairs.txt', pre_dir, p_top_peak, p_top_gene) %>% read.table(header = T)
peak_gene_pairs <- get_top_peak_gene_pairs(ct_obj,
                                           gene_top=p_top_gene, peak_top=p_top_peak,
                                           distance = 5e+5,
                                           gene_assay = 'RNA', 
                                           peak_assay = pn)
res_df <- scMultiMap(ct_obj, peak_gene_pairs,
                     bsample = ifelse(adjust_for_batch, , NULL),
                     gene_assay = 'RNA', # name of the gene assay
                     peak_assay = pn) 
    
write.table(res_df, sprintf('%s_proposed_observed_results%s%s.txt', output_dir, fn_suffix, batch_suffix))
print(sprintf('%s_proposed_observed_results%s%s.txt', output_dir, fn_suffix, batch_suffix))

if(adjust_for_batch){
    saveRDS(batch_inds_list,
            sprintf('%s_proposed_observed_batch%s%s.rds', output_dir, fn_suffix, batch_suffix))
    print(sprintf('%s_proposed_observed_batch%s%s.rds', output_dir, fn_suffix, batch_suffix))
}

# -
# for trio analysis
# a customized implementation of CS-CORE for evaluating gene-gene correlation
# reference: https://www.nature.com/articles/s41467-023-40503-7
# -

if(run_TF_gene){
    source('../method_implementation/moment_regressions.R')
    library(Signac)
    print('evaluate TF-gene co-expressions')

    # extract data from the seurat object
    rna_seq_depths <- colSums(ct_obj[['RNA']]@counts)
    peak_seq_depths <- colSums(ct_obj[[pn]]@counts)
    unique_genes <- unique(peak_gene_pairs$gene)
    unique_peaks <- unique(peak_gene_pairs$peak)
    
    ## generated in preprocessing/generate_motif_object.R
    motif.object <- readRDS(sprintf('../preprocessing/output/%s/motif.object.rds', study))
    full_TF_names <- ConvertMotifID(motif.object, id=colnames(motif.object@data))
    # focus on TFs that have expression levels profiled in scRNA-seq
    TF_names <- full_TF_names[full_TF_names %in% rownames(ct_obj[['RNA']]@counts)] 
    genes_high_TF <- unique(c(TF_names, unique_genes))
    # run CS-CORE for gene gene co-expression between TFs and target genes
    gene_mar_pars <- mom_marginal_adj_for_batch(ct_obj[['RNA']]@counts[genes_high_TF, ] %>% as.matrix %>% t, 
                                                rna_seq_depths, batch=batch_mat, irls=T)
    TF_inds <- match(TF_names, genes_high_TF)
    non_TF_inds <- which(!genes_high_TF %in% TF_names)
    TF_target_p <- TF_target_coexp <- matrix(nrow = length(TF_names), ncol = length(genes_high_TF))
    rownames(TF_target_p) <- rownames(TF_target_coexp) <- TF_names
    colnames(TF_target_p) <- colnames(TF_target_coexp) <- genes_high_TF
    for(i in 1:length(TF_names)){
        tmp <- mom_covariance(gene_mar_pars$X_centered[, TF_inds[i]], 
                                gene_mar_pars$X_centered,#[, non_TF_inds], 
                                gene_mar_pars$w[, TF_inds[i]],
                                gene_mar_pars$w,#[, non_TF_inds],
                                rna_seq_depths,
                                rna_seq_depths,
                                gene_mar_pars$X_var[, TF_inds[i]],
                                gene_mar_pars$X_var)#[, non_TF_inds])
        TF_target_p[i,] <- tmp$test_stat
        TF_target_coexp[i,] <- tmp$covar / sqrt(gene_mar_pars$sigma_sq[TF_inds[i]] * gene_mar_pars$sigma_sq) # [non_TF_inds]
    }
    TF_target_p <- 2 * pnorm(abs(TF_target_p), lower.tail = F)
    fn <- sprintf('%s_TF_gene_coexp%s%s.rds', output_dir, fn_suffix, batch_suffix)
    print(fn)
    saveRDS(list(pval = TF_target_p, coexp = TF_target_coexp, 
                 sparsity = rowSums(ct_obj[['RNA']]@counts == 0) / ncol(ct_obj[['RNA']]@counts)), fn)
#}

#if(run_TF_peak){
#    library(Signac)
#    print('evaluate TF-peak co-expressions')
#    ## generated in preprocessing/generate_motif_object.R
#    motif.object <- readRDS('output/brain_CG/motif.object.rds')
#    full_TF_names <- ConvertMotifID(motif.object, id=colnames(motif.object@data))
#    # focus on TFs that have expression levels profiled in scRNA-seq
#    TF_names <- full_TF_names[full_TF_names %in% rownames(ct_obj[['RNA']]@counts)] 

    motif_peaks_sub <- motif.object@data[rownames(motif.object@data) %in% peak_gene_pairs$peak,]
    # for each TF, evaluate the association with peaks that harbor its binding motif
    TF_peak_list <- list()
    for(i in 1:length(TF_names)){
        peaks_w_binding_motifs <- names(which(motif_peaks_sub[,full_TF_names == TF_names[i]] != 0))
        results_peaks_w_motif_inds <- match(peaks_w_binding_motifs, unique_peaks) # only extract unique peaks
        # results$peak %in% peaks_w_binding_motifs # there could be repeated peaks with different genes
        tmp <- mom_covariance(gene_mar_pars$X_centered[, TF_inds[i]], 
                              peak_mar_pars$X_centered[, results_peaks_w_motif_inds],
                              gene_mar_pars$w[, TF_inds[i]],
                              peak_mar_pars$w[, results_peaks_w_motif_inds],
                              rna_seq_depths,
                              peak_seq_depths,
                              gene_mar_pars$X_var[, TF_inds[i]],
                              peak_mar_pars$X_var[, results_peaks_w_motif_inds])
        tmp_df <- data.frame(test_stat = tmp$test_stat,
                            covar = tmp$covar,
			    coexp = tmp$covar / sqrt(gene_mar_pars$sigma_sq[TF_inds[i]] * peak_mar_pars$sigma_sq[results_peaks_w_motif_inds]))
        tmp_df$pval <- 2 * pnorm(abs(tmp_df$test_stat), lower.tail = F)
        rownames(tmp_df) <- peaks_w_binding_motifs
        TF_peak_list[[TF_names[i]]] <- tmp_df
    }
    fn <- sprintf('%s_TF_peak_coexp%s%s.rds', output_dir, fn_suffix, batch_suffix)
    print(fn)
    saveRDS(TF_peak_list, fn)
}
